### DISCUSSION TITLE
Possibility to create a branch to track current stable release?

### DISCUSSION THREAD

**Discussion by p0358** [curl/curl#11213] [Ideas]
So there is currently the `master` branch aka the current development tree, and there are also tags for stable versions such as `curl-8_1_1`. I would like to bring into the consideration the idea of creating a branch like `stable` or `release`, which would always be set to track the current stable release, equivalent to the latest stable release tag.

The reason I ask for this is that it'd make it easy to put in `branch = stable` into `.gitmodules`, to always pull in the latest stable version instead of development when updating submodules (since dev one can be broken sometimes), and it's tedious to always manually update to tags after checking them, since you cannot set a submodule to track a tag in the `.gitmodules` file.

There are many projects out there that do something like this, so I hope this is a reasonable idea that could be put to life, thanks in advance

**Comment by bagder** [comment]
If you want the latest stable, then I recommend using the latest release tag. Then you just need to update that whenever there is a new tag and you are ready to jump to the next new release.

I think there are many benefits with keeping the simplicity of a simple linear history and working continuously on master. I don't think it is worth slightly complicating the life of the developers just to be able to provide a "stable branch".

**Reply by p0358** [reply]
Eh, just push the stable branch to particular commit at the same time release is made, it shouldn't be too impacting on the developers. That still keeps continuous history, I'm not asking to forgo the master branch or anything. It's just that the stable branch would be set to some commits before current master, just like a tag.

Problems with tags are that they're persistent, and need to be updated manually, plus you cannot track a tag in `,gitmodules`, meaning you need to fiddle with manually updating the commit the submodule points at every time. Or well, do as some projects do with just pointing it at master, but the point is to use latest stable release, rather than particular random master commit that might end up being broken.

> I don't think it is worth slightly complicating the life of the developers just to be able to provide a "stable branch".

Eh, I think it's a minimal complication for libcurl when a release is made (I assume there's some CI or script template when that happens), at the cost of biggest complication for anyone updating the submodules.

And for curl it should really be just as simple as `git tag curl-8_1_1` and then just `git push origin stable`...

**Reply by jzakrzewski** [reply]
You probably misunderstood each other.
What @p0358 means is a "release" branch, not "stable". Something that simply follows the release tags, and always points to the same commit as the latest release tag. 
I wouldn't use something like that for my own projects just to avoid unexpected updates.

**Reply by bagder** [reply]
So basically a tag that is always set on (moved to) the latest release tag?

**Reply by jzakrzewski** [reply]
That's how I understand this request, yes.

**Reply by p0358** [reply]
> So basically a tag that is always set on (moved to) the latest release tag?

Yes, but implemented as a branch that's automatically fast-forwarded to the latest release tag once it's out (because physically updating an existing git tag instead comes with its big set of issues).

Then someone can always pull such branch after fetching, and they're always on the latest release tag, without checking what that tag currently is manually

**Reply by p0358** [reply]
> I wouldn't use something like that for my own projects just to avoid unexpected updates.

Well, gitmodules current commit still has to be updated manually, so it's still an explicit action, just that tracking a branch upstream allows the updates to be performed for all deps automatically without checking what commit was the version we want to update to. Then the project can be recompiled and manually tested that it still works etc

**Reply by p0358** [reply]
Oh, but speaking of unexpected updates avoidance, for such case there could be release branches such as `curl-8.x` which would work the same way as proposed `release` next to them, except they'd stop getting updated with next major version (as with semver major versions are expected to potentially have breaking changes)

**Comment by bagder** [comment]
This can can be done to figure out the git tag for the current release:
~~~
$ curl -s https://curl.se/info | grep "^git tag" | cut -d" " -f3
curl-8_1_2

**Reply by jay** [reply]
another way:
~~~sh
curl -fLsS https://api.github.com/repos/curl/curl/releases/latest | jq -j ".tag_name | select(type == \"string\")"
~~~

**Comment by dfandrich** [comment]
My understanding of the request is that a tag won't do—it needs to be a branch
that's fast forwarded on every release.

**Reply by p0358** [reply]
Yes exactly 🙏

**Comment by p0358** [comment]
@bagder any updates on this? would make things way easier

**Reply by bagder** [reply]
No updates.

I'm not convinced this is in our general interest. We already have a fair amount of problems with people who think what is tagged in git are our releases and I fear a "release branch" would risk strengthen that illusion.

Our releases are signed tarballs that are *generated* from the git repo using the specific tag.

**Reply by p0358** [reply]
> We already have a fair amount of problems with people who think what is tagged in git are our releases and I fear a "release branch" would risk strengthen that illusion.

I don't believe having a branch for latest tag would change anything in that regard though (wouldn't make situation better or worse).

One problem with current situation is that if we run (to update all submodules):

> git submodule update --remote --merge

Then we get an error:

> fatal: Unable to find refs/remotes/origin/curl-8_2_1 revision in submodule path 'deps/curl'

Cause .gitmodules doesn't support tags. But if we were to point it at master, then curl would inevitably be automatically updated with other submodules.

So this causes complication in which all submodules now have to be explicitly updated manually, because curl's case breaks updating stuff all-at-once. Writing scripts also isn't a nice solution, because it clutters the workflow and isn't cross-OS compatible...
